NPE in raft.removeNode(mcc.msg.NodeID)

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x4671397]

goroutine 4154 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc820463020)
        /Users/tschottdorf/go/src/github.com/coreos/etcd/raft/multinode.go:253 +0x14e7
created by github.com/coreos/etcd/raft.StartMultiNode
        /Users/tschottdorf/go/src/github.com/coreos/etcd/raft/multinode.go:56 +0x30d
```

offending code:
```go
			case pb.ConfChangeRemoveNode:
				group.raft.removeNode(mcc.msg.NodeID)
```

in (not raft-pertinent) WIP freshly rebased on top of master at time of writing. Can't reproduce.

--- a/multiraft/clock.go
+++ b/multiraft/clock.go
@@ -61,8 +61,21 @@ func (m *manualTicker) Chan() <-chan time.Time {
 	return m.ch
 }
 
+// Tick sends a tick to MultiRaft, blocking until MultiRaft is ready
+// to receive it. Use this when it is important to send a specific
+// number of ticks.
 func (m *manualTicker) Tick() {
 	m.ch <- time.Time{}
+}
+
+// NonBlockingTick tries to send a tick to MultiRaft, silently
+// dropping it if MultiRaft is not listening. Use this when sending
+// ticks from a background thread that may race with shutdown.
+func (m *manualTicker) NonBlockingTick() {
+	select {
+	case m.ch <- time.Time{}:
+	default:
+	}
 }
 
 func (m *manualTicker) Close() { /* do nothing */ }

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -227,7 +227,7 @@ func (s *state) fanoutHeartbeat(req *RaftMessageRequest) {
 	// heartbeat. If we get a heartbeat from a node we don't know, it
 	// must think we are a follower of some group, so we need to respond
 	// so it can activate the recovery process.
-	s.sendMessage(noGroup,
+	s.sendMessage(nil,
 		raftpb.Message{
 			From: uint64(s.nodeID),
 			To:   req.Message.From,
@@ -371,6 +371,8 @@ type proposal struct {
 
 // group represents the state of a consensus group.
 type group struct {
+	id proto.RangeID
+
 	// committedTerm is the term of the most recently committed entry.
 	committedTerm uint64
 
@@ -604,7 +606,7 @@ func (s *state) coalescedHeartbeat() {
 		if log.V(6) {
 			log.Infof("node %v: triggering coalesced heartbeat to node %v", s.nodeID, nodeID)
 		}
-		s.sendMessage(noGroup,
+		s.sendMessage(nil,
 			raftpb.Message{
 				From: uint64(s.nodeID),
 				To:   uint64(nodeID),
@@ -621,19 +623,9 @@ func (s *state) stop() {
 	s.MultiRaft.Transport.Stop(s.nodeID)
 }
 
-// addNode creates a node and registers the given groupIDs for that
-// node. If the node already exists and possible some of the groups
-// are already registered, only the missing groups will be added in.
-func (s *state) addNode(nodeID proto.RaftNodeID, groupIDs ...proto.RangeID) error {
-	for _, groupID := range groupIDs {
-		if groupID == noGroup {
-			panic(fmt.Sprintf("attempt to add dummy group to node %d", nodeID))
-		}
-		if _, ok := s.groups[groupID]; !ok {
-			return util.Errorf("can not add invalid group %d to node %d",
-				groupID, nodeID)
-		}
-	}
+// addNode creates a node and registers the given group (if not nil)
+// for that node.
+func (s *state) addNode(nodeID proto.RaftNodeID, g *group) error {
 	newNode, ok := s.nodes[nodeID]
 	if !ok {
 		s.nodes[nodeID] = &node{
@@ -643,21 +635,21 @@ func (s *state) addNode(nodeID proto.RaftNodeID, groupIDs ...proto.RangeID) erro
 		}
 		newNode = s.nodes[nodeID]
 	}
-	for _, groupID := range groupIDs {
-		newNode.registerGroup(groupID)
-		g := s.groups[groupID]
+
+	if g != nil {
+		newNode.registerGroup(g.id)
 		g.nodeIDs = append(g.nodeIDs, nodeID)
 	}
 	return nil
 }
 
 // removeNode removes a node from a group.
-func (s *state) removeNode(nodeID proto.RaftNodeID, groupID proto.RangeID) error {
+func (s *state) removeNode(nodeID proto.RaftNodeID, g *group) error {
 	node, ok := s.nodes[nodeID]
 	if !ok {
 		return util.Errorf("cannot remove unknown node %s", nodeID)
 	}
-	g := s.groups[groupID]
+
 	for i := range g.nodeIDs {
 		if g.nodeIDs[i] == nodeID {
 			g.nodeIDs[i] = g.nodeIDs[len(g.nodeIDs)-1]
@@ -666,7 +658,7 @@ func (s *state) removeNode(nodeID proto.RaftNodeID, groupID proto.RangeID) error
 		}
 	}
 	// TODO(bdarnell): when a node has no more groups, remove it.
-	node.unregisterGroup(groupID)
+	node.unregisterGroup(g.id)
 
 	// Cancel any outstanding proposals.
 	if nodeID == s.nodeID {
@@ -713,12 +705,14 @@ func (s *state) createGroup(groupID proto.RangeID) error {
 	if err := s.multiNode.CreateGroup(uint64(groupID), raftCfg, nil); err != nil {
 		return err
 	}
-	s.groups[groupID] = &group{
+	g := &group{
+		id:      groupID,
 		pending: map[string]*proposal{},
 	}
+	s.groups[groupID] = g
 
 	for _, nodeID := range cs.Nodes {
-		if err := s.addNode(proto.RaftNodeID(nodeID), groupID); err != nil {
+		if err := s.addNode(proto.RaftNodeID(nodeID), g); err != nil {
 			return err
 		}
 	}
@@ -909,9 +903,9 @@ func (s *state) processCommittedEntry(groupID proto.RangeID, g *group, entry raf
 						// TODO(bdarnell): dedupe by keeping a record of recently-applied commandIDs
 						switch cc.Type {
 						case raftpb.ConfChangeAddNode:
-							err = s.addNode(proto.RaftNodeID(cc.NodeID), proto.RangeID(groupID))
+							err = s.addNode(proto.RaftNodeID(cc.NodeID), g)
 						case raftpb.ConfChangeRemoveNode:
-							err = s.removeNode(proto.RaftNodeID(cc.NodeID), proto.RangeID(groupID))
+							err = s.removeNode(proto.RaftNodeID(cc.NodeID), g)
 						case raftpb.ConfChangeUpdateNode:
 							// Updates don't concern multiraft, they are simply passed through.
 						}
@@ -944,23 +938,21 @@ func (s *state) processCommittedEntry(groupID proto.RangeID, g *group, entry raf
 
 // sendMessage sends a raft message on the given group. Coalesced heartbeats
 // address nodes, not groups; they will use the noGroup constant as groupID.
-func (s *state) sendMessage(groupID proto.RangeID, msg raftpb.Message) {
+func (s *state) sendMessage(g *group, msg raftpb.Message) {
 	if log.V(6) {
 		log.Infof("node %v sending message %.200s to %v", s.nodeID,
 			raft.DescribeMessage(msg, s.EntryFormatter), msg.To)
 	}
 	nodeID := proto.RaftNodeID(msg.To)
+	groupID := noGroup
+	if g != nil {
+		groupID = g.id
+	}
 	if _, ok := s.nodes[nodeID]; !ok {
 		if log.V(4) {
 			log.Infof("node %v: connecting to new node %v", s.nodeID, nodeID)
 		}
-		var err error
-		if groupID != noGroup {
-			err = s.addNode(nodeID, groupID)
-		} else {
-			err = s.addNode(nodeID)
-		}
-		if err != nil {
+		if err := s.addNode(nodeID, g); err != nil {
 			log.Errorf("node %v: error adding group %v to node %v: %v",
 				s.nodeID, groupID, nodeID, err)
 		}
@@ -1053,7 +1045,7 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 			for _, nodeID := range ready.Snapshot.Metadata.ConfState.Nodes {
 				// TODO(bdarnell): if we had any information that predated this snapshot
 				// we must remove those nodes.
-				if err := s.addNode(proto.RaftNodeID(nodeID), raftGroupID); err != nil {
+				if err := s.addNode(proto.RaftNodeID(nodeID), g); err != nil {
 					log.Errorf("node %v: error adding node %v", s.nodeID, nodeID)
 				}
 			}
@@ -1076,7 +1068,7 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 						s.nodeID, msg.To)
 				}
 			default:
-				s.sendMessage(raftGroupID, msg)
+				s.sendMessage(g, msg)
 			}
 		}
 	}

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -651,6 +651,33 @@ func (s *state) addNode(nodeID proto.RaftNodeID, groupIDs ...proto.RangeID) erro
 	return nil
 }
 
+// removeNode removes a node from a group.
+func (s *state) removeNode(nodeID proto.RaftNodeID, groupID proto.RangeID) error {
+	node, ok := s.nodes[nodeID]
+	if !ok {
+		return util.Errorf("cannot remove unknown node %s", nodeID)
+	}
+	g := s.groups[groupID]
+	for i := range g.nodeIDs {
+		if g.nodeIDs[i] == nodeID {
+			g.nodeIDs[i] = g.nodeIDs[len(g.nodeIDs)-1]
+			g.nodeIDs = g.nodeIDs[:len(g.nodeIDs)-1]
+			break
+		}
+	}
+	// TODO(bdarnell): when a node has no more groups, remove it.
+	node.unregisterGroup(groupID)
+
+	// Cancel any outstanding proposals.
+	if nodeID == s.nodeID {
+		for _, prop := range g.pending {
+			s.removePending(g, prop, ErrGroupDeleted)
+		}
+	}
+
+	return nil
+}
+
 func (s *state) createGroup(groupID proto.RangeID) error {
 	if _, ok := s.groups[groupID]; ok {
 		return nil
@@ -758,6 +785,20 @@ func (s *state) propose(p *proposal) {
 		s.removePending(nil /* group */, p, ErrGroupDeleted)
 		return
 	}
+
+	found := false
+	for _, nodeID := range g.nodeIDs {
+		if nodeID == s.nodeID {
+			found = true
+			break
+		}
+	}
+	if !found {
+		// If we are not a member of the group, don't allow any proposals.
+		s.removePending(nil, p, ErrGroupDeleted)
+		return
+	}
+
 	if log.V(3) {
 		log.Infof("group %d: new proposal %x", p.groupID, p.commandID)
 	}
@@ -870,7 +911,7 @@ func (s *state) processCommittedEntry(groupID proto.RangeID, g *group, entry raf
 						case raftpb.ConfChangeAddNode:
 							err = s.addNode(proto.RaftNodeID(cc.NodeID), proto.RangeID(groupID))
 						case raftpb.ConfChangeRemoveNode:
-							// TODO(bdarnell): support removing nodes; fix double-application of initial entries
+							err = s.removeNode(proto.RaftNodeID(cc.NodeID), proto.RangeID(groupID))
 						case raftpb.ConfChangeUpdateNode:
 							// Updates don't concern multiraft, they are simply passed through.
 						}

--- a/multiraft/multiraft_test.go
+++ b/multiraft/multiraft_test.go
@@ -432,7 +432,7 @@ func TestRemoveLeader(t *testing.T) {
 				return
 			case <-ticker.C:
 				for _, t := range cluster.tickers {
-					t.Tick()
+					t.NonBlockingTick()
 				}
 			}
 		}

--- a/multiraft/multiraft_test.go
+++ b/multiraft/multiraft_test.go
@@ -392,6 +392,78 @@ func TestMembershipChange(t *testing.T) {
 		}*/
 }
 
+// TestRemoveLeader ensures that a group will recover if a node is
+// removed from the group while it is leader. Since visibility into
+// the raft state is limited, we create a three-node group in a
+// six-node cluster. This group is migrated one node at a time from
+// the first three nodes to the last three. In the process the initial
+// leader must have removed itself.
+func TestRemoveLeader(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	stopper := stop.NewStopper()
+	const clusterSize = 6
+	const groupSize = 3
+	cluster := newTestCluster(nil, clusterSize, stopper, t)
+	defer stopper.Stop()
+
+	// Consume and apply the membership change events.
+	for i := 0; i < clusterSize; i++ {
+		go func(i int) {
+			for {
+				if e, ok := <-cluster.events[i].MembershipChangeCommitted; ok {
+					e.Callback(nil)
+				} else {
+					return
+				}
+			}
+		}(i)
+	}
+
+	// Tick all the clocks in the background to ensure that all the
+	// necessary elections are triggered.
+	// TODO(bdarnell): newTestCluster should have an option to use a
+	// real clock instead of a manual one.
+	stopper.RunWorker(func() {
+		ticker := time.NewTicker(10 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-stopper.ShouldStop():
+				return
+			case <-ticker.C:
+				for _, t := range cluster.tickers {
+					t.Tick()
+				}
+			}
+		}
+	})
+
+	// Create a group with three members.
+	groupID := proto.RangeID(1)
+	cluster.createGroup(groupID, 0, groupSize)
+
+	// Move the group one node at a time from the first three nodes to
+	// the last three. In the process, we necessarily remove the leader
+	// and trigger at least one new election among the new nodes.
+	for i := 0; i < groupSize; i++ {
+		log.Infof("adding node %d", i+groupSize)
+		ch := cluster.nodes[i].ChangeGroupMembership(groupID, makeCommandID(),
+			raftpb.ConfChangeAddNode,
+			cluster.nodes[i+groupSize].nodeID, nil)
+		if err := <-ch; err != nil {
+			t.Fatal(err)
+		}
+
+		log.Infof("removing node %d", i)
+		ch = cluster.nodes[i].ChangeGroupMembership(groupID, makeCommandID(),
+			raftpb.ConfChangeRemoveNode,
+			cluster.nodes[i].nodeID, nil)
+		if err := <-ch; err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
 func TestRapidMembershipChange(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	stopper := stop.NewStopper()

